library/kotlin/test/io/envoyproxy/envoymobile/GRPCStreamEmitterTest.kt

```diff
@@ -0,0 +1,109 @@
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.io.ByteArrayOutputStream
+import java.nio.ByteBuffer
+
+class GRPCStreamEmitterTest {
+
+  // TODO: Problems with nhaarman/mockito-kotlin https://github.com/lyft/envoy-mobile/issues/504
+  // This is a total hack to get something to work
+  private lateinit var emitter: StreamEmitter
+
+  private val dataOutputStream = ByteArrayOutputStream()
+  private val closeOutputStream = ByteArrayOutputStream()
+  private var isCloseCalled = false
+
+  @Before
+  fun setup() {
+    emitter = object : StreamEmitter {
+
+      override fun cancel() {
+        TODO("not implemented")
+      }
+
+      override fun sendData(byteBuffer: ByteBuffer): StreamEmitter {
+        dataOutputStream.write(byteBuffer.array())
+        return this
+      }
+
+      override fun sendMetadata(metadata: Map<String, List<String>>): StreamEmitter {
+        TODO("not implemented")
+      }
+
+      override fun close(byteBuffer: ByteBuffer) {
+        isCloseCalled = true
+        closeOutputStream.write(byteBuffer.array())
+      }
+
+      override fun close(trailers: Map<String, List<String>>) {
+        TODO("not implemented")
+      }
+    }
+  }
+
+  @After
+  fun teardown() {
+    dataOutputStream.reset()
+    closeOutputStream.reset()
+    isCloseCalled = false
+  }
+
+  @Test
+  fun `heading and data is sent on send data`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray()).hasSize(payload.size + MESSAGE_HEADING_OFFSET)
+  }
+
+  @Test
+  fun `compression flag is set on the first bit of the heading`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray()[0]).isEqualTo(0)
+  }
+
+  @Test
+  fun `message length is set on the 1-4 bytes of the heading`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(ByteBuffer.wrap(dataOutputStream.toByteArray().sliceArray(1..4)).int).isEqualTo(payload.size)
+  }
+
+  @Test
+  fun `message is sent after the heading`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray().sliceArray(MESSAGE_HEADING_OFFSET until dataOutputStream.size()).toString(Charsets.UTF_8)).isEqualTo("data")
+
+  }
+
+  @Test
+  fun `close is called with empty data frame`() {
```

Comment:
@junr03 we should add a test like this on iOS as part of https://github.com/lyft/envoy-mobile/pull/500

Reply:
https://github.com/lyft/envoy-mobile/pull/509